examples/sde_gan.py

```diff
@@ -0,0 +1,523 @@
+    # condition is wrong then it's pretty hard to learn the rest of the SDE correctly, so it's good to start off with an
+    # initial condition that's as close to being right as possible.
+    # (You could probably also do something like train to match just the initial condition first, and then train to
+    # match the rest of the SDE, but I've not tried that.)
+    ###################
+    y0_flat = ys[0].view(-1)
+    y0_not_nan = y0_flat.masked_select(~torch.isnan(y0_flat))
+    ys = (ys - y0_not_nan.mean()) / y0_not_nan.std()
+
+    ###################
+    # As discussed, time must be included as a channel for the discriminator.
+    ###################
+    ys = torch.cat([ts.unsqueeze(0).unsqueeze(-1).expand(dataset_size, t_size, 1),
+                    ys.transpose(0, 1)], dim=2)
+    # shape (dataset_size=1000, t_size=100, 1 + data_size=3)
+
+    return ts, ys
+
+
+###################
+# Now do normal GAN training, and plot the results.
+# We need some standard GAN tricks (e.g. stochastic weight averaging) to get this working.
+###################
+
+def train_generator(ts, batch_size, generator, discriminator, generator_optimiser, discriminator_optimiser):
+    generated_samples = generator(ts, batch_size)
+    generated_score = discriminator(generated_samples)
+
+    generated_score.backward()
+    generator_optimiser.step()
+    generator_optimiser.zero_grad()
+    discriminator_optimiser.zero_grad()
+
+
+def train_discriminator(ts, batch_size, real_samples, generator, discriminator, discriminator_optimiser, device,
+                        gp_coeff):
+    with torch.no_grad():
+        generated_samples = generator(ts, batch_size)
+    generated_score = discriminator(generated_samples)
+
+    real_samples = real_samples.to(device)
+    real_score = discriminator(real_samples)
+
+    penalty = gradient_penalty(generated_samples, real_samples, discriminator)
+    loss = generated_score - real_score
+    (gp_coeff * penalty - loss).backward()
+    discriminator_optimiser.step()
+    discriminator_optimiser.zero_grad()
+
+
+def evaluate_loss(ts, batch_size, dataloader, generator, discriminator, device):
+    with torch.no_grad():
+        total_samples = 0
+        total_loss = 0
+        for real_samples, in dataloader:
+            generated_samples = generator(ts, batch_size)
+            generated_score = discriminator(generated_samples)
+
+            real_samples = real_samples.to(device)
+            real_score = discriminator(real_samples)
+
+            loss = generated_score - real_score
+
+            batch_size = real_samples.size(0)
+            total_samples += batch_size
+            total_loss += loss.item() * batch_size
+    return total_loss / total_samples
+
+
+def main():
+    # Architectural hyperparameters. These are quite small for illustrative purposes.
+    initial_noise_size = 5  # How many noise dimensions to sample at the start of the SDE.
```

Comment:
I'm not a fan of argparse in example code: increased boilerplate; reduced readability.

Reply:
If you want a CLI but are worried about boilerplate maybe https://github.com/google/python-fire might fit the bill?
(Haven't tried it myself so YMMV)

Reply:
I have been using fire for quite a while, though I was relunctant to include it for other examples, as it required another dependency. 

Thinking about it now, it's not that big a hassle, and I have modified the other examples like so. 

In general not a huge deal though.

Reply:
A side comment is that I would still recommend using some method so that the script accepts command line args directly. Or perhaps at the very least, make args like `initial_noise_size` parameters of the function so they can be tweaked if people wanted to.